src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -337,13 +340,17 @@ private void EventLoop()
                         }
                     }
 
-                    // HandleEvents queues to the ThreadPool; so we queue the next read 
-                    // after all the events have been picked up by the ThreadPool (but not necessarily complete)
-                    // This makes it more likely there are events and the batches are larger (less syscalls)
-                    // as well as not overloading the system with events at a higher rate than they can actually be 
-                    // processed, which breaks any back pressure on the sockets.
-                    ThreadPool.UnsafeQueueUserWorkItem((mres) => ((ManualResetEventSlim)mres).Set(), _batchProcessed);
-                    _batchProcessed.Wait();
+                    if (numEvents > BatchThreshold)
+                    {
+                        // Batch received; HandleEvents queues to the ThreadPool; so we queue the next read 
+                        // after all the events have been picked up by the ThreadPool (but not necessarily complete)
+                        // This makes it more likely there are events and the batches are larger (less syscalls)
+                        // as well as not overloading the system with events at a higher rate than they can actually be 
+                        // processed, which breaks any back pressure on the sockets.
+                        _batchProcessed.Reset();
+                        ThreadPool.UnsafeQueueUserWorkItem(s_releaseBatch, _batchProcessed);
```

Comment:
I actually wasn't referring to a delegate allocation... I was referring to the QueueUserWorkItemCallback object allocation.  The way to avoid that would be, for example, to have a ManualResetEventSlim-derived type that also implements IThreadPoolWorkItem, and just requeue that one object repeatedly.

Reply:
Ah! I'm clearly off my game :)